src/clj/rems/workflow/permissions.clj

```diff
@@ -0,0 +1,166 @@
+  "Sets role specific permissions for the application.
+
+   In `permission-map`, the key is the role (a keyword), and the value
+   is a list of permissions to set for that role (also keywords).
+   The permissions may represent commands that the user is allowed to run,
+   or they may be used to specify whether the user can see all events and
+   comments from the reviewers (e.g. `:see-everything`)."
+  [application permission-map]
+  (reduce (fn [application [role permissions]]
+            (assert (keyword? role))
+            (assoc-in application [::role-permissions role] (set permissions)))
+          application
+          permission-map))
+
+(deftest test-set-role-permissions
+  (testing "adding"
+    (is (= {::role-permissions {:role #{:foo :bar}}}
+           (-> {}
+               (set-role-permissions {:role [:foo :bar]})))))
+  (testing "updating"
+    (is (= {::role-permissions {:role #{:gazonk}}}
+           (-> {}
+               (set-role-permissions {:role [:foo :bar]})
+               (set-role-permissions {:role [:gazonk]})))))
+  (testing "removing"
+    (is (= {::role-permissions {:role #{}}}
+           (-> {}
+               (set-role-permissions {:role [:foo :bar]})
+               (set-role-permissions {:role []}))))
+    (is (= {::role-permissions {:role #{}}}
+           (-> {}
+               (set-role-permissions {:role [:foo :bar]})
+               (set-role-permissions {:role nil})))))
+
+  (testing "can set permissions for multiple roles"
+    (is (= {::role-permissions {:role-1 #{:foo}
+                                :role-2 #{:bar}}}
+           (-> {}
+               (set-role-permissions {:role-1 [:foo]
+                                      :role-2 [:bar]})))))
+  (testing "does not alter unrelated roles"
+    (is (= {::role-permissions {:unrelated #{:foo}
+                                :role #{:gazonk}}}
+           (-> {}
+               (set-role-permissions {:unrelated [:foo]
+                                      :role [:bar]})
+               (set-role-permissions {:role [:gazonk]}))))))
+
+(defn remove-permission-from-all [application permission]
+  (let [roles (keys (::role-permissions application))]
+    (reduce (fn [application role]
+              (update-in application [::role-permissions role] disj permission))
+            application
+            roles)))
+
+(deftest test-remove-permission-from-all
+  (testing "removes the permission from all roles"
+    (is (= {::role-permissions {:role-1 #{}
+                                :role-2 #{}}}
+           (-> {}
+               (set-role-permissions {:role-1 [:foo]
+                                      :role-2 [:foo]})
+               (remove-permission-from-all :foo)))))
+  (testing "leaves unrelated permissions unchanged"
+    (is (= {::role-permissions {:role #{:bar}}}
+           (-> {}
+               (set-role-permissions {:role [:foo :bar]})
+               (remove-permission-from-all :foo))))))
+
+(defn user-permissions
```

Comment:
This could be written for example

```
(defn user-permissions
  "Returns the specified user's permissions to this application.
   Union of all role specific permissions."
  [application user]
  (->> (user-roles application user)
       (mapcat (fn [role]
              (get-in application [::role-permissions role])))
       (set)))
```